File: backend/app/schemas/ask.py
"""
Pydantic schemas for the /ask API (Phases 9–10).

POST /ask is the standalone orchestration endpoint (pre-conversation form):
it runs the full RAG pipeline and streams the answer as SSE — superseded by
the conversation endpoint in Phase 11, retained for testing and the
evaluation harness (roadmap Phase 9 §APIs).

Phase 10: the terminal ``done`` payload now carries the resolved, validated
``citations[]`` (FE §12 Citation contract) plus the post-validation answer
text — citations are emitted only AFTER generation + validation complete,
never mid-stream (Backend §37).

The scope request shape is shared with POST /search (same VersionScope
resolution semantics — current-document/selected/knowledge-base map onto
the same three kinds).
"""
from __future__ import annotations

from datetime import date
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field

from app.schemas.search import SearchScopeRequest


# ── Request ───────────────────────────────────────────────────────────────────

class AskHistoryMessage(BaseModel):
    """One prior conversation turn (bounded — the server keeps the last few)."""

    role: Literal["user", "assistant"]
    content: Annotated[
        str,
        Field(min_length=1, max_length=4000, description="The turn's text."),
    ]


class AskRequest(BaseModel):
    """POST /ask request body."""

    question: Annotated[
        str,
        Field(
            min_length=1,
            max_length=2000,
            description="The user's question (the ORIGINAL words — the only "
                        "text ever answered or persisted; any retrieval rewrite "
                        "is an internal artifact).",
            examples=["What is the approval process?"],
        ),
    ]
    scope: SearchScopeRequest | None = Field(
        default=None,
        description=(
            "Optional document-scope narrowing (advisory hints in the "
            "question never EXPAND it — the authoritative scope is this "
            "selection intersected with the caller's permissions).  "
            "Omit for the entire knowledge base."
        ),
    )
    history: list[AskHistoryMessage] | None = Field(
        default=None,
        description=(
            "Recent conversation turns (max 6) enabling follow-up question "
            "rewriting.  No persistence happens in Phase 9 — the caller "
            "supplies the context."
        ),
    )
    top_k: Annotated[
        int | None,
        Field(
            default=None,
            ge=1,
            le=20,
            description=(
                "Maximum sources considered for context (defaults to the "
                "server setting — the documented 5–8 band)."
            ),
        ),
    ] = None


# ── Response (SSE payloads) ───────────────────────────────────────────────────

class AskSourceItem(BaseModel):
    """One source in the ``sources`` SSE event."""

    index: int = Field(description="1-based SOURCE label the answer cites as [N].")
    chunk_id: str = Field(description="UUID of the source chunk (backend-resolved).")
    document_id: str = Field(description="UUID of the source document.")
    document_version_id: str = Field(description="UUID of the searched version.")
    document_name: str = Field(description="Human-readable document name.")
    page_id: str = Field(description="UUID of the page the chunk starts on (citation anchor).")
    page_number: int = Field(description="1-indexed page the chunk starts on.")
    section_title: str | None = Field(
        default=None, description="Section heading (null for unstructured documents)."
    )
    relevance: float = Field(ge=0.0, le=1.0, description="Presentation score [0, 1].")
    snippet: str = Field(description="First 500 characters of the source content.")


class AskCitationItem(BaseModel):
    """One resolved, validated citation (the ``done`` payload's ``citations[]``).

    The FE §12 Citation contract: everything the badge/popover/navigation
    needs — the numbered index matching the inline [N] marker, the exact
    quoted span with char offsets (real source text, never model output —
    Backend §35), its surrounding context for the preview popover, and the
    document/version/page navigation target.
    """

    index: int = Field(description="The [N] ordinal as it appears inline in the answer.")
    chunk_id: str = Field(description="UUID of the exact chunk grounding this citation.")
    document_id: str = Field(description="UUID of the source document.")
    document_version_id: str = Field(description="UUID of the cited version.")
    document_name: str = Field(description="Human-readable document name.")
    version_number: int | None = Field(
        default=None, description="Version number of the cited version (badge display)."
    )
    effective_date: date | None = Field(
        default=None, description="The cited version's effective date, when set."
    )
    page_id: str = Field(description="UUID of the cited page (navigation anchor).")
    page: int = Field(description="1-indexed page number for direct viewer navigation.")
    section: str | None = Field(
        default=None, description="Denormalized section title/path (null if unstructured)."
    )
    text: str = Field(
        description=(
            "quoted_text — the exact source span from the ACTUAL chunk content "
            "(never model-paraphrased text mislabeled as a quote)."
        )
    )
    char_start: int | None = Field(
        default=None,
        description="Offset of the quoted span within the chunk content (highlight overlay).",
    )
    char_end: int | None = Field(default=None, description="End offset (exclusive).")
    context_before: str = Field(
        default="", description="Chunk text preceding the quoted span (preview popover)."
    )
    context_after: str = Field(
        default="", description="Chunk text following the quoted span (preview popover)."
    )
    relevance: float = Field(
        ge=0.0, le=1.0, description="Retrieval relevance of the cited chunk [0, 1]."
    )


class AskDonePayload(BaseModel):
    """The terminal ``done`` SSE event payload."""

    groundedness: Literal["grounded", "partial", "ungrounded"] = Field(
        description=(
            "'grounded' when an evidence-constrained, validated answer was "
            "generated; 'partial' when some claims were stripped or could not "
            "be verified; 'ungrounded' is the explicit insufficient-evidence "
            "outcome — a SUCCESS state, not an error (Backend §48)."
        )
    )
    message: str | None = Field(
        default=None,
        description="User-facing note (set on the ungrounded path).",
    )
    message_id: str | None = Field(
        default=None,
        description=(
            "UUID of the persisted assistant message (written atomically with "
            "its citations — Backend §50).  Null only if persistence failed."
        ),
    )
    answer: str = Field(
        default="",
        description=(
            "The FINAL post-validation answer text.  May differ from the "
            "accumulated token stream: uncited/unsupported claims were "
            "stripped and invalid reference markers removed — render THIS."
        ),
    )
    citations: list[AskCitationItem] = Field(
        default_factory=list,
        description="Resolved, validated citations (never streamed mid-generation).",
    )
    model: str | None = Field(default=None, description="The generating model.")
    prompt_tokens: int = Field(default=0)
    completion_tokens: int = Field(default=0)
    intent: str = Field(default="QUESTION", description="Analyzer intent class.")
    topic: str | None = Field(default=None, description="Analyzer topic label (analytics).")
    used_rewrite: bool = Field(
        default=False,
        description="True when a drift-guarded standalone rewrite drove retrieval.",
    )
    used_reranker: bool = Field(default=False)
    regenerated: bool = Field(
        default=False,
        description=(
            "True when the shipped answer is a citation-emphasis regeneration "
            "(one bounded retry — Backend §36)."
        ),
    )
    stripped_claims: int = Field(
        default=0, description="Claims removed by validation (uncited/unsupported)."
    )
    entailment_checks: int = Field(
        default=0, description="Entailment verification calls made for this answer."
    )
    injection_attempt: bool = Field(
        default=False,
        description=(
            "Phase 16: True when the canary sentinel fired — a document "
            "attempted instruction hijack; contaminated sentence(s) were "
            "stripped. FE shows a subtle security notice."
        ),
    )
    latency_ms: dict[str, Any] = Field(
        default_factory=dict,
        description=(
            "Per-stage latency_ms (analyzer/rewrite/retrieval/context/"
            "generation/citations)."
        ),
    )

File: frontend/src/lib/api/ask.ts
/**
 * Ask AI API — SSE streaming client (Phases 9–10).
 *
 * POST /ask returns a text/event-stream: `token`* → `sources` → `done`
 * (plus `error` for provider failures).  Axios buffers responses, so the
 * stream is consumed with fetch + ReadableStream; the Authorization header
 * comes from the same in-memory token store the axios client uses.
 *
 * Insufficient evidence is a SUCCESS-shaped stream (sources: [], done with
 * groundedness: 'ungrounded') — the FE §6.6 "no grounded answer found"
 * state, never an error banner.
 *
 * Phase 10: the done payload now carries resolved `citations[]` (FE §12
 * Citation contract) and the post-validation `answer` text — citations are
 * emitted only AFTER generation + validation complete, never mid-stream.
 */

import { API_BASE_URL } from './client'
import { tokenStore } from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

/** One source attached to the answer (the context blocks behind it). */
export interface AskSource {
  index: number
  chunk_id: string
  document_id: string
  document_version_id: string
  document_name: string
  page_id: string
  page_number: number
  section_title: string | null
  relevance: number
  snippet: string
}

/**
 * One resolved, validated citation (FE §12 Citation contract): the numbered
 * marker matching the inline [N] in the answer, the exact quoted span with
 * char offsets (REAL source text — never model paraphrase), its surrounding
 * context, and the document/version/page navigation target.
 */
export interface AskCitation {
  index: number
  chunk_id: string
  document_id: string
  document_version_id: string
  document_name: string
  version_number: number | null
  effective_date: string | null
  page_id: string
  page: number
  section: string | null
  text: string
  char_start: number | null
  char_end: number | null
  context_before: string
  context_after: string
  relevance: number
}

export interface AskScope {
  document_ids?: string[]
  collection_ids?: string[]
}

export interface AskHistoryTurn {
  role: 'user' | 'assistant'
  content: string
}

export interface AskRequestBody {
  question: string
  scope?: AskScope
  history?: AskHistoryTurn[]
}

export type AskStreamEvent =
  | { type: 'token'; text: string }
  | { type: 'sources'; sources: AskSource[] }
  | {
      type: 'done'
      groundedness: 'grounded' | 'partial' | 'ungrounded'
      message: string | null
      message_id: string | null
      /** Post-validation answer text — render THIS, not the raw token stream. */
      answer: string
      citations: AskCitation[]
      model: string | null
      prompt_tokens: number
      completion_tokens: number
      intent: string
      topic: string | null
      used_rewrite: boolean
      used_reranker: boolean
      regenerated: boolean
      stripped_claims: number
      entailment_checks: number
      latency_ms: Record<string, number>
    }
  | { type: 'error'; code: string; message: string }

export interface AskStreamHandlers {
  onEvent: (event: AskStreamEvent) => void
}

// ─── SSE parsing ──────────────────────────────────────────────────────────────

/**
 * Parse one SSE block (`event: X\ndata: {...}`) into a typed AskStreamEvent.
 * Unknown event types are ignored (forward compatibility).
 */
function parseSseBlock(block: string): AskStreamEvent | null {
  let eventName = ''
  const dataLines: string[] = []
  for (const line of block.split('\n')) {
    if (line.startsWith('event:')) eventName = line.slice(6).trim()
    else if (line.startsWith('data:')) dataLines.push(line.slice(5).trim())
  }
  if (!eventName) return null
  const data = dataLines.length ? JSON.parse(dataLines.join('\n')) : {}
  switch (eventName) {
    case 'token':
      return { type: 'token', text: String(data.text ?? '') }
    case 'sources':
      return { type: 'sources', sources: data.sources ?? [] }
    case 'done':
      return { type: 'done', ...data }
    case 'error':
      return {
        type: 'error',
        code: String(data.code ?? 'UNKNOWN_ERROR'),
        message: String(data.message ?? 'The AI service is unavailable.'),
      }
    default:
      return null
  }
}

// ─── Stream request ───────────────────────────────────────────────────────────

/** Build the request scope from the ScopeSelector state (empty selection → all). */
export function toAskScope(
  mode: 'all' | 'selected',
  selectedIds: string[],
): AskScope | undefined {
  if (mode === 'selected' && selectedIds.length > 0) {
    return { document_ids: selectedIds }
  }
  return undefined
}

/**
 * POST /ask and dispatch each parsed SSE event to `onEvent`.
 *
 * The caller controls cancellation via `signal` (the FE §6.6 stop-generating
 * control; server-side cancellation lands in Phase 11).  Resolves when the
 * stream ends; throws ApiError-shaped errors only for non-200 responses
 * before the stream starts (401/422 etc.).
 */
export async function streamAsk(
  body: AskRequestBody,
  handlers: AskStreamHandlers,
  signal?: AbortSignal,
): Promise<void> {
  const token = tokenStore.get()
  const response = await fetch(`${API_BASE_URL}/ask`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'text/event-stream',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(body),
    signal,
  })

  if (!response.ok) {
    let code = 'UNKNOWN_ERROR'
    let message = 'An unexpected error occurred.'
    try {
      const payload = await response.json()
      if (payload?.error) {
        code = payload.error.code ?? code
        message = payload.error.message ?? message
      }
    } catch {
      // non-JSON error body — keep the defaults
    }
    throw Object.assign(new Error(message), { code, httpStatus: response.status })
  }

  if (!response.body) {
    throw Object.assign(new Error('Streaming is not supported by this browser.'), {
      code: 'STREAM_UNSUPPORTED',
    })
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  const dispatch = (block: string) => {
    const trimmed = block.trim()
    if (!trimmed) return
    try {
      const event = parseSseBlock(trimmed)
      if (event) handlers.onEvent(event)
    } catch {
      // Malformed block — skip it rather than killing the stream
    }
  }

  for (;;) {
    const { done, value } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })
    let separator = buffer.indexOf('\n\n')
    while (separator !== -1) {
      dispatch(buffer.slice(0, separator))
      buffer = buffer.slice(separator + 2)
      separator = buffer.indexOf('\n\n')
    }
  }
  // Flush any final unterminated block (defensive — the server always ends with \n\n)
  dispatch(buffer)
}
